deepinterpolation/generator_collection.py

```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
-            self.raw_data_file = self.json_data["train_path"]
-        else:
-            self.raw_data_file = self.json_data["movie_path"]
-
+        self.cache_data = self.json_data.get("cache_data", False)
+        self.raw_data_file = self.json_data.get(
+            "train_path", self.json_data.get("movie_path"))
+        self.gpu_cache_full = self.json_data.get("gpu_cache_full", False)
+        self.normalize_cache = self.json_data.get("normalize_cache", False)
         self.batch_size = self.json_data["batch_size"]
-        movie_obj_point = h5py.File(self.raw_data_file, "r")
-        raw_data = movie_obj_point["data"]
+        self.movie_statistics_sample_size = \
+            self.json_data.get("movie_statistics_sample_size")
 
-        self.total_frame_per_movie = int(raw_data.shape[0])
+        self.total_frame_per_movie = int(self.movie_data.shape[0])
 
         self._update_end_frame(self.total_frame_per_movie)
         self._calculate_list_samples(self.total_frame_per_movie)
 
-        average_nb_samples = np.min([int(raw_data.shape[0]), 100])
-
-        # For backward compatibility
-        if "cache_data" in self.json_data.keys():
-            self.cache_data = self.json_data["cache_data"]
-        else:
-            self.cache_data = False
-
-        if self.cache_data:
-            print('Caching hdf5 file... \n')
-            if self.end_frame > 0:
-                self.raw_data = raw_data[0:np.min([self.total_frame_per_movie,
-                    self.end_frame+self.post_frame+self.pre_post_omission+1]), :, :]
-            else:
-                self.raw_data = raw_data[:, :, :]
-            local_data = self.raw_data[0:average_nb_samples, :, :].flatten()
-        else:
-            local_data = raw_data[0:average_nb_samples, :, :].flatten()
-            self.cache_data = False
-
+        average_nb_samples = np.min(
+            [int(self.movie_data.shape[0]), self.movie_statistics_sample_size])
+        local_data = self.movie_data[:average_nb_samples].flatten()
         local_data = local_data.astype("float32")
 
-        self.local_mean = np.mean(local_data)
-        self.local_std = np.std(local_data)
+        self.local_mean = local_data.mean()
+        self.local_std = local_data.std()
+
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
+        if self._movie_data is None:
+            with h5py.File(self.raw_data_file, "r") as movie_obj:
+                end_ind = self.end_frame+self.post_frame+self.pre_post_omission+1
+                total_frame_per_movie = movie_obj['data'].shape[0]
+                if self.end_frame > 0 and  total_frame_per_movie > end_ind:
+                    self._movie_data = movie_obj['data'][:end_ind]
+                else:
+                    self._movie_data = movie_obj['data'][()]
+            if self.gpu_cache_full:
+                logger.info("Caching full movie onto GPU")
+                self._data_tensor = tf.convert_to_tensor(
```

Comment:
I propose to set `movie_data` to the tensor. Then the return type will be correct.

Reply:
Yeah, it's a lot slower for processing each batch on the CPU. Might be the slicing operation is less efficient with tf than np.